### ISSUE TITLE
Feature request - save every n steps

### ISSUE THREAD

**Issue by cyber-meow** [kohya-ss/sd-scripts#239]
For big datasets an epoch can take a lot of hours. It would be much more practical if checkpoints can be saved every n steps. After looking at the code I think the modification should be rather straightforward.

**Comment by KohakuBlueleaf** [comment]
Also need this

**Comment by dill-shower** [comment]
It's good idea

**Comment by armandoalonso** [comment]
this would be awesome

**Comment by cyber-meow** [comment]
For reference, this is what I use locally, added after `if global_step >= args.max_train_steps: break`
```
     if args.save_every_n_steps is not None and global_step % args.save_every_n_steps == 0:
        model_name =  "step" if args.output_name is None else args.output_name + "-step"
        ckpt_name = f"{model_name}-{global_step}" '.' + args.save_model_as
        ckpt_file = os.path.join(args.output_dir, ckpt_name)
        metadata["ss_training_finished_at"] = str(time.time())
        print(f"saving checkpoint: {ckpt_file}")
        unwrap_model(network).save_weights(ckpt_file, save_dtype, None if args.no_metadata else metadata)
```
Of course along with
```
  parser.add_argument("--save_every_n_steps", type=int, default=None)
```

**Comment by swfsql** [comment]
I'm also in favor of this since I'm experimenting training with lots of data with my potato GPU.  
I believe this could also be related to a streamability of the (data) -> (cached data) -> (training) procedure. For example, one thing that could help would be to have a training's cached data be stored to disk, and to reference that cached data in a later training.

This is kinda funny, but I just got a, unfortunate OOM error (I thought my settings would fit it, goddamn!)
```
caching latents.
100%|██████████| 124992/124992 [4:31:23<00:00,  7.68it/s]
running training / 学習開始
epoch 1/20
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB
```

**Comment by cyber-meow** [comment]
Indeed. Caching latents can also takes nearly one hour in some of my datasets. I know that Kohya has some scripts that store cached latents (https://github.com/kohya-ss/sd-scripts/blob/main/finetune/prepare_buckets_latents.py) but I am not sure if it is directly compatible with all the training scripts.

Back until a while a go some tweaks were still needed for that. Someone sent me this
https://github.com/kohya-ss/sd-scripts/compare/main...breakcore2:sd-scripts:finetune_like_dreambooth but I have never looked into it.

**Comment by GeneralAwareness** [comment]
Far better to save every steps than every epochs for me at least with the testing I have done, and I did read this was the findings of others as well.

We need this.
